lib/_http_incoming.js

```diff
@@ -114,19 +114,25 @@ IncomingMessage.prototype._read = function _read(n) {
     readStart(this.socket);
 };
 
-
 // It's possible that the socket will be destroyed, and removed from
 // any messages, before ever calling this.  In that case, just skip
 // it, since something else is destroying this connection anyway.
-IncomingMessage.prototype.destroy = function destroy(error) {
-  // TODO(ronag): Implement in terms of _destroy
-  this.destroyed = true;
-  if (this.socket)
-    this.socket.destroy(error);
-  return this;
+IncomingMessage.prototype._destroy = function _destroy(err, cb) {
+  if (this.req && this.socket && !this.readableEnded) {
+    // On the client, if destroy is called before end, destroy the underlying
+    // socket, if present.
+    this.socket.destroy(err);
+  } else if (
+    this.socket &&
+    this.socket.parser &&
+    this.socket.parser.incoming === this) {
+    // On the server, if destroy is called while still parsing, manually call
+    // emitCloseNT on the Stream.Readable side because the event won't be
+    // emitted from the server.
+    cb(err);
+  }
```

Comment:
callback must always be invoked or you get corrupted state, there is not much point in autodestroy if we don't implement it properly